Sources/PointFree/Auth.swift

```diff
@@ -95,34 +95,34 @@ private func writeGitHubSessionCookieMiddleware(
 private func authTokenMiddleware(
   _ conn: Conn<StatusLineOpen, (code: String, redirect: String?)>
   )
-  -> IO<Conn<ResponseEnded, Data>> {
+  -> IO<Conn<ResponseEnded, Either<Error, Data>>> {
 
     return AppEnvironment.current.fetchAuthToken(conn.data.code)
       .flatMap { token in
         AppEnvironment.current.fetchGitHubUser(token)
           .map { user in GitHubUserEnvelope(accessToken: token, gitHubUser: user) }
       }
       .run
-      .flatMap { githubUserEnvelope in
-        switch githubUserEnvelope {
-
-        case .left:
-          return conn
-            |> redirect(to: path(to: .secretHome))
-
+      .flatMap { gitHubUserEnvelope in
+
+        switch gitHubUserEnvelope {
+        case let .left(error):
+          return conn.map(const(Either<Error, Data>.left(error)))
+            |> writeStatus(.internalServerError)
+            >-> end
+            >>> map(map(pure))
```

Comment:
This is the ugly part. Our middleware combinators always expect to end on `Data`, so both of these branches have to move the monad index forward and re-wrap the result in `Either`.